Exercises.AsyncAwait/UrlSaver.cs

```diff
@@ -29,37 +37,73 @@ public async Task<bool> TryGetGetUrlsFromStorageAsync()
             return _urls.Count > 0;
         }
 
-        public void GetDataFromUrlAsync(int downloadsQuantity)
+        public async Task<int> GetDataFromUrlAsync(int downloadsQuantity)
         {
-            var queue = new Queue<string>();
+            var t = _taskCompletionSource.Task;
+
             foreach (var url in _urls)
             {
-                queue.Enqueue(url.Key);
+                _queue.Enqueue(url.Key);
             }
+            _quantumOfStatusLine = 100m / _queue.Count;
+            _logger.Write($"In queue {_queue.Count} downloads");
+
             var tasksInTimeQuantity = downloadsQuantity >= _urls.Count
                 ? _urls.Count
                 : downloadsQuantity;
 
-            var tasksCompleted = 0;
-
-            while (tasksCompleted != _urls.Count)
+            for (int i = 0; i < tasksInTimeQuantity; i++)
             {
-                Task.Run(() =>
-               {
-                   await GetContentFromUrlAsync(queue.Dequeue());
+                TaskRun();
+                _workingTasks++;
+            }
 
-               });
+            return await t;
+        }
+
+        private void TaskRun()
+        {
+            var newTask = GetContentFromUrlAsync(_queue.Dequeue());
+            newTask.ContinueWith((e) =>
+            {
+                _workingTasks--;
+                if (_queue.Count != 0)
+                {
+                    TaskRun();
+                    _workingTasks++;
+                }
+                else if (_workingTasks == 0)
+                {
+                    _taskCompletionSource.SetResult(1);
+                }
+            });
+        }
 
+        public async Task SaveContentToStorageAsync()
+        {
+            foreach (var url in _urls)
+            {
+                await _storage.SaveContentToStorageAsync(url.Key, url.Value);
             }
+        }
 
+
+        private async Task GetContentFromUrlAsync(string url)
+        {
+            _logger.Write("Start new download");
```

Comment:
If you write `throw new Exception();` in that line, your program will never know (catch) it